networking_cisco/plugins/cisco/cfg_agent/service_helpers/routing_svc_helper_aci.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2014 Cisco Systems, Inc.  All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from oslo_log import log as logging
+from oslo_utils import excutils
+import pprint as pp
+
+from neutron.common import constants as l3_constants
+
+from networking_cisco.plugins.cisco.cfg_agent import cfg_exceptions
+from networking_cisco.plugins.cisco.cfg_agent.service_helpers import (
+    routing_svc_helper as helper)
+from networking_cisco.plugins.cisco.common import (cisco_constants as
+                                                   c_constants)
+from networking_cisco.plugins.cisco.extensions import routerrole
+
+ROUTER_ROLE_ATTR = routerrole.ROUTER_ROLE_ATTR
+LOG = logging.getLogger(__name__)
+
+
+class RoutingServiceHelperAci(helper.RoutingServiceHelper):
+
+    def __init__(self, host, conf, cfg_agent):
+        super(RoutingServiceHelperAci, self).__init__(
+            host, conf, cfg_agent)
+        self._router_ids_by_vrf = {}
+
+    def _process_router(self, ri):
+        """Process a router, apply latest configuration and update router_info.
+
+
+        :param ri : RouterInfo object of the router being processed.
+        :return:None
+        :raises: networking_cisco.plugins.cisco.cfg_agent.cfg_exceptions.
+        DriverException if the configuration operation fails.
+        """
+        try:
+            ex_gw_port = ri.router.get('gw_port')
+            gateway_set = ex_gw_port and not ri.ex_gw_port
+            gateway_cleared = not ex_gw_port and ri.ex_gw_port
+            # Once the gateway is set, then we know which VRF
+            # this router belongs to. Keep track of it in our
+            # lists of routers, organized as a dictionary by
+            # VRF name
+            if gateway_set:
+                self._add_rid_to_vrf_list(ri)
+
+            curr_interfaces = ri.router.get(l3_constants.INTERFACE_KEY, [])
+            LOG.debug("++ curr_interfaces = %s" %
+                      (pp.pformat([p['id'] for p in curr_interfaces])))
+
+            existing_int_ids = set([p['id'] for p in ri.internal_ports])
```

Comment:
if u will do the refactoring then my comment here is moot as those codes will be re-used.